substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSections.java

```diff
@@ -0,0 +1,2344 @@
+    // private static final byte DW_CFA_def_cfa_register = 0xd;
+    private static final byte DW_CFA_def_cfa_offset = 0xe;
+
+    private ELFMachine elfMachine;
+    private DwarfStrSectionImpl dwarfStrSection;
+    private DwarfAbbrevSectionImpl dwarfAbbrevSection;
+    private DwarfInfoSectionImpl dwarfInfoSection;
+    private DwarfARangesSectionImpl dwarfARangesSection;
+    private DwarfLineSectionImpl dwarfLineSection;
+    private DwarfFrameSectionImpl dwarfFameSection;
+
+    public DwarfSections(ELFMachine elfMachine) {
+        this.elfMachine = elfMachine;
+        dwarfStrSection = new DwarfStrSectionImpl();
+        dwarfAbbrevSection = new DwarfAbbrevSectionImpl();
+        dwarfInfoSection = new DwarfInfoSectionImpl();
+        dwarfARangesSection = new DwarfARangesSectionImpl();
+        dwarfLineSection = new DwarfLineSectionImpl();
+        dwarfFameSection = (elfMachine == ELFMachine.AArch64
+                        ? new DwarfFrameSectionImplAArch64()
+                        : new DwarfFrameSectionImplX86_64());
+    }
+
+    public DwarfStrSectionImpl getStrSectionImpl() {
+        return dwarfStrSection;
+    }
+
+    public DwarfAbbrevSectionImpl getAbbrevSectionImpl() {
+        return dwarfAbbrevSection;
+    }
+
+    public DwarfFrameSectionImpl getFrameSectionImpl() {
+        return dwarfFameSection;
+    }
+
+    public DwarfInfoSectionImpl getInfoSectionImpl() {
+        return dwarfInfoSection;
+    }
+
+    public DwarfARangesSectionImpl getARangesSectionImpl() {
+        return dwarfARangesSection;
+    }
+
+    public DwarfLineSectionImpl getLineSectionImpl() {
+        return dwarfLineSection;
+    }
+
+    public ELFMachine getElfMachine() {
+        return elfMachine;
+    }
+
+    /**
+     * a scratch buffer used during computation of a section's size.
+     */
+    protected static final byte[] scratch = new byte[10];
+
+    /**
+     * a table listing all known strings, some of
+     * which may be marked for insertion into the
+     * debug_str section.
+     */
+    private StringTable stringTable = new StringTable();
+
+    /**
+     * list detailing all dirs in which files are found to reside
+     * either as part of substrate/compiler or user code.
+     */
+    private LinkedList<DirEntry> dirs = new LinkedList<>();
+    /**
+     * index of already seen dirs.
+     */
+    private Map<String, DirEntry> dirsIndex = new HashMap<>();
+
+    // The obvious traversal structure for debug records is:
+    //
+    // 1) by top level compiled method (primary Range) ordered by ascending address
+    // 2) by inlined method (sub range) within top level method ordered by ascending address
+    //
+    // this ensures that all debug records are generated in increasing address order
+
+    /**
+     * a list recording details of all primary ranges included in
+     * this file sorted by ascending address range.
+     */
+    private LinkedList<PrimaryEntry> primaryEntries = new LinkedList<>();
```

Comment:
afaics you only ever add to this list without every using it afterwards

Reply:
Yes, I was hedging on the idea that it might be necessary at some point to iterate over (outer) compiled methods in address order. At present all iteration is by methods within each owning class (using the list of ClassEntry) which is effectively the same. I have removed this list and updated the comments regarding iteration options.